File: javascript/src/repositories/lease.ts
import { z } from "zod";
import type { CallProc } from "./types.js";
import { DeductionRowSchema } from "./deduction.js";
import type { DeductionRow } from "./deduction.js";
import { pgBoolean, safeParse } from "./_shared.js";

export const LeaseRowSchema = z
  .object({
    lease_id: z.string().optional(),
    user_id: z.string().optional(),
    amount: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    available: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    reserved: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    billing_mode: z.string().optional(),
    expires_at: z.string().optional(),
    error: z.string().optional(),
  })
  .passthrough();

export const ReleaseRowSchema = z
  .object({
    released: pgBoolean.nullable().optional(),
    reason: z.string().nullable().optional(),
  })
  .passthrough();

export type LeaseRow = z.infer<typeof LeaseRowSchema>;
export type ReleaseRow = z.infer<typeof ReleaseRowSchema>;

/** Repository for lease lifecycle operations (admission control). */
export class LeaseRepository {
  constructor(private callproc: CallProc) {}

  /** Atomically acquire a lease (hold) — admission control. */
  async createLease(params: {
    userId: string;
    amount: string;
    operationType: string;
    billingMode: string;
    floor: string;
    maxConcurrent: number | null;
    ttlSeconds: number;
    model: string | null;
    overdraftFloor: string | null;
    metadata: string;
    periodStart: string | null;
    feature: string | null;
    featureMaxCalls: number | null;
    featureOnExceed: string | null;
    featurePeriodStart: string | null;
    featurePeriodEnd: string | null;
  }): Promise<LeaseRow> {
    const rows = await this.callproc("create_lease", [
      params.userId,
      params.amount,
      params.operationType,
      params.billingMode,
      params.floor,
      params.maxConcurrent,
      params.ttlSeconds,
      params.model,
      params.overdraftFloor,
      params.metadata,
      params.periodStart,
      params.feature,
      params.featureMaxCalls,
      params.featureOnExceed,
      params.featurePeriodStart,
      params.featurePeriodEnd,
    ]);
    return safeParse(LeaseRowSchema, rows?.[0] ?? {}, "LeaseRepository.createLease");
  }

  /** Charge the actual cost against a lease and mark it settled. */
  async settleLease(params: {
    userId: string;
    leaseId: string;
    amount: string;
    idempotencyKey: string | null;
    minBalance: string;
    model: string | null;
    metadata: string;
    skipAllowance: boolean;
    periodStart: string | null;
    feature: string | null;
    featureMaxCalls: number | null;
    featureOnExceed: string | null;
    featurePeriodStart: string | null;
    featurePeriodEnd: string | null;
  }): Promise<DeductionRow> {
    const rows = await this.callproc("settle_lease", [
      params.userId,
      params.leaseId,
      params.amount,
      params.idempotencyKey,
      params.minBalance,
      params.model,
      params.metadata,
      params.skipAllowance,
      params.periodStart,
      params.feature,
      params.featureMaxCalls,
      params.featureOnExceed,
      params.featurePeriodStart,
      params.featurePeriodEnd,
    ]);
    return safeParse(DeductionRowSchema, rows?.[0] ?? {}, "LeaseRepository.settleLease");
  }

  /** Release a lease without charging — idempotent. */
  async releaseLease(userId: string, leaseId: string): Promise<ReleaseRow> {
    const rows = await this.callproc("release_lease", [userId, leaseId]);
    return safeParse(ReleaseRowSchema, rows?.[0] ?? {}, "LeaseRepository.releaseLease");
  }

  /** Extend an active lease's TTL. */
  async renewLease(userId: string, leaseId: string, ttlSeconds: number): Promise<LeaseRow> {
    const rows = await this.callproc("renew_lease", [userId, leaseId, ttlSeconds]);
    return safeParse(LeaseRowSchema, rows?.[0] ?? {}, "LeaseRepository.renewLease");
  }
}

File: python/src/bursar/repositories/lease.py
from __future__ import annotations

from bursar.repositories._types import DbQuery
from bursar.repositories._utils import validate_amount, validate_non_empty
from bursar.repositories.schemas import CreateLeaseParams, DeductionRow, LeaseRow, ReleaseRow, SettleLeaseParams


class LeaseRepository:
    """Repository for lease lifecycle operations.

    All methods call Postgres RPCs via the callproc function.
    Returns None when the RPC returns no rows (no result).
    Returns typed Pydantic models for successful results.
    """

    def __init__(self, callproc: DbQuery) -> None:
        self._callproc = callproc

    def create_lease(self, params: CreateLeaseParams) -> LeaseRow | None:
        """Create a credit lease (reservation) for a user.

        Args:
            params: CreateLeaseParams with user_id, amount, operation_type,
                billing_mode, floor, max_concurrent, ttl_seconds, model,
                overdraft_floor, metadata, period_start, and optional
                feature limit fields.

        Returns:
            LeaseRow if created, None if the RPC returned no rows.
        """
        validate_non_empty(params.user_id, "user_id")
        validate_amount(params.amount, "amount")
        rows = self._callproc(
            "create_lease",
            [
                params.user_id,
                params.amount,
                params.operation_type,
                params.billing_mode,
                params.floor,
                params.max_concurrent,
                params.ttl_seconds,
                params.model,
                params.overdraft_floor,
                params.metadata,
                params.period_start,
                params.feature,
                params.feature_max_calls,
                params.feature_action,
                params.feature_period_start,
                params.feature_period_end,
            ],
        )
        if not rows:
            return None
        return LeaseRow.model_validate(rows[0])

    def settle_lease(self, params: SettleLeaseParams) -> DeductionRow | None:
        """Settle a lease by deducting the actual amount used.

        Args:
            params: SettleLeaseParams with user_id, lease_id, amount,
                idempotency_key, min_balance, model, metadata,
                skip_allowance, period_start, and optional feature limit fields.

        Returns:
            DeductionRow if settled, None if the RPC returned no rows.
        """
        validate_non_empty(params.user_id, "user_id")
        validate_non_empty(params.lease_id, "lease_id")
        rows = self._callproc(
            "settle_lease",
            [
                params.user_id,
                params.lease_id,
                params.amount,
                params.idempotency_key,
                params.min_balance,
                params.model,
                params.metadata,
                params.skip_allowance,
                params.period_start,
                params.feature,
                params.feature_max_calls,
                params.feature_action,
                params.feature_period_start,
                params.feature_period_end,
            ],
        )
        if not rows:
            return None
        return DeductionRow.model_validate(rows[0])

    def release_lease(self, user_id: str, lease_id: str) -> ReleaseRow | None:
        """Release a lease without deducting credits (cancels the reservation).

        Args:
            user_id: The user ID.
            lease_id: The lease ID to release.

        Returns:
            ReleaseRow if released, None if the lease was not found.
        """
        validate_non_empty(user_id, "user_id")
        validate_non_empty(lease_id, "lease_id")
        rows = self._callproc("release_lease", [user_id, lease_id])
        if not rows:
            return None
        return ReleaseRow.model_validate(rows[0])

    def renew_lease(self, user_id: str, lease_id: str, ttl_seconds: int) -> LeaseRow | None:
        """Extend the TTL of an existing lease.

        Args:
            user_id: The user ID.
            lease_id: The lease ID to renew.
            ttl_seconds: The new TTL in seconds from now.

        Returns:
            LeaseRow with updated expiry, or None if the lease was not found.
        """
        validate_non_empty(user_id, "user_id")
        validate_non_empty(lease_id, "lease_id")
        rows = self._callproc("renew_lease", [user_id, lease_id, ttl_seconds])
        if not rows:
            return None
        return LeaseRow.model_validate(rows[0])
